algebra/euclidean_domain.lean

```diff
@@ -0,0 +1,324 @@
+/- 
+Copyright (c) 2018 Louis Carlin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Louis Carlin
+
+Euclidean domains and Euclidean algorithm (extended to come)
+A lot is based on pre-existing code in mathlib for natural number gcds
+-/
+import tactic.ring
+
+universe u
+
+class euclidean_domain (α : Type u) extends integral_domain α :=
+( quotient : α → α → α )
+( remainder : α → α → α )
+( witness : ∀ a b, (quotient a b) * b + (remainder a b) = a ) -- This could be changed to the same order as int.mod_add_div. We normally write qb+r rather than r + qb though.
+(valuation : α → ℕ)
+(val_mod : ∀ a b, b = 0 ∨  valuation (remainder a b) <  valuation b)
+(val_mul : ∀ a b, b = 0 ∨ valuation a ≤ valuation (a*b))
+/-
+val_mul is often not a required in definitions of a euclidean domain since given the other properties we can show there is a (noncomputable) euclidean domain α with the property val_mul.
+So potentially this definition could be split into two different ones (euclidean_domain_weak and euclidean_domain_strong) with a noncomputable function from weak to strong.
+I've currently divided the lemmas into strong and weak depending on whether they require val_mul or not.
+-/
+
+class decidable_euclidean_domain (α : Type u) extends euclidean_domain α:=
+(decidable_eq_zero : ∀ a : α, decidable (a = 0))
+/-
+We use this for basically everything.
+It might be worth making it part of the original definition?
+-/
+
+namespace euclidean_domain
+variable {α : Type u}
+
+instance decidable_eq_zero  [ed : decidable_euclidean_domain α] (a : α) : decidable (a = 0) :=
+ decidable_euclidean_domain.decidable_eq_zero a
+
+instance euclidean_domain_has_div [euclidean_domain α] : has_div α := { 
+    div := quotient
+}
+
+instance euclidean_domain_has_mod [euclidean_domain α] : has_mod α := {
+    mod := remainder
+}
+
+instance ed_has_sizeof [euclidean_domain α] : has_sizeof α := {
+    sizeof := λ x : α, valuation x,
+}
+
+definition gcd_decreasing [euclidean_domain α] (a b : α) (w : a ≠ 0) : has_well_founded.r (b % a) a := 
+begin
+    cases val_mod b a,
+    { contradiction },
+    { exact h }
+end
+
+def gcd [decidable_euclidean_domain α] : α → α → α
+| a b := if a_zero : a = 0 then b
+         else have h : has_well_founded.r (b % a) a := gcd_decreasing a b a_zero,
+              gcd (b%a) a
+
+/- weak lemmas -/
+
+@[simp] lemma mod_zero [ed : euclidean_domain α] (a : α)  : a % 0 = a :=
+begin
+    have := euclidean_domain.witness a 0,
+    simp at this,
+    exact this,
+end
```

Comment:
Is there a way of doing this more quickly? I think the philosophy is that if the lemma has a tiny bit of content then there's an argument for readability, but if it's a mathematical triviality then I think the mathlib guys prefer an incomprehensible one-liner. Oh -- how about

`@[simp] lemma mod_zero (a : α) : a % 0 = a := by simpa using euclidean_domain.quotient_mul_add_remainder_eq a 0`

I think that's pretty much the same as your proof -- if you're using VS Code then type "simpa" into it and hover your mouse over it to see the docs.